File: web/javascript/driverSelenium/basePage.js
const { By, until } = require('selenium-webdriver');

class BasePage {
    /**
     * Constructor for BasePage.
     * @param {WebDriver} driver - The Selenium WebDriver object.
     */
    constructor(driver) {
        this.driver = driver;
        this.defaultTimeout = 30000; // Default timeout in milliseconds
    }

    /**
     * Clicks on an element.
     * @param {By} locator - The Selenium By locator for the element.
     */
    async doClick(locator) {
        const element = await this.driver.wait(until.elementLocated(locator), this.defaultTimeout);
        await this.driver.wait(until.elementIsVisible(element), this.defaultTimeout);
        await element.click();
    }

    /**
     * Sends keystrokes to an element.
     * @param {By} locator - The Selenium By locator for the element.
     * @param {string} text - The text to type.
     */
    async sendKeys(locator, text) {
        const element = await this.driver.wait(until.elementLocated(locator), this.defaultTimeout);
        await element.sendKeys(text);
    }

    /**
     * Navigates to a specified URL.
     * @param {string} url - The URL to navigate to.
     */
    async navigate(url) {
        await this.driver.get(url);
    }

    /**
     * Wait for URL to contain.
     * @param {string} text - The URL to contain this text.
     */
    async waitForUrlToContain(text) {
        await this.driver.wait(until.urlContains(text), this.defaultTimeout);
    }
}

module.exports = BasePage;

File: web/python/cucumber/driver_selenium/base_page.py
from selenium.webdriver.remote.webdriver import WebDriver
from selenium.webdriver.remote.webelement import WebElement
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.action_chains import ActionChains
from selenium.common.exceptions import TimeoutException

class BasePage:
    """
    BasePage class for common page actions using Selenium.
    """

    def __init__(self, driver: WebDriver, default_timeout: int = 30):
        self.driver = driver
        self.default_timeout = default_timeout  # Default timeout in seconds

    def do_click(self, element: WebElement, timeout=None):
        timeout = timeout or self.default_timeout
        WebDriverWait(self.driver, timeout).until(EC.visibility_of(element))
        ActionChains(self.driver).move_to_element(element).click().perform()

    def send_keys(self, element: WebElement, text, timeout=None):
        timeout = timeout or self.default_timeout
        WebDriverWait(self.driver, timeout).until(EC.visibility_of(element))
        element.clear()
        element.send_keys(text)

    def navigate(self, url):
        self.driver.get(url)

    def wait_for_url_to_contain(self, text, timeout=None):
        timeout = timeout or self.default_timeout
        try:
            WebDriverWait(self.driver, timeout).until(EC.url_contains(text))
        except TimeoutException:
            raise Exception(f"URL did not contain the text '{text}' within {timeout} seconds")

